vital/tests/test_metadata.cxx

```diff
@@ -0,0 +1,122 @@
+ *    and/or other materials provided with the distribution.
+ *
+ *  * Neither name of Kitware, Inc. nor the names of any contributors may be used
+ *    to endorse or promote products derived from this software without specific
+ *    prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <vital/types/metadata.h>
+#include <vital/types/metadata_traits.h>
+
+#include <gtest/gtest.h>
+
+#include <memory>
+#include <stdint.h>
+
+using namespace ::kwiver::vital;
+
+// ----------------------------------------------------------------------------
+int
+main( int argc, char** argv )
+{
+  ::testing::InitGoogleTest( &argc, argv );
+  return RUN_ALL_TESTS();
+}
+
+// ----------------------------------------------------------------------------
+TEST ( metadata, typed_metadata )
+{
+  // create item
+  auto* tmds = NEW_METADATA_ITEM( VITAL_META_METADATA_ORIGIN, std::string( "item data" ) );
+  auto* tmdd = new typed_metadata< VITAL_META_PLATFORM_HEADING_ANGLE, double >(
+    std::string( "test double item" ), 3.14159 );
+  auto* tmdi = new typed_metadata< VITAL_META_UNIX_TIMESTAMP, uint64_t >(
+    std::string( "test uint item" ), uint64_t( 314159 ) );
+
+  // test API
+  EXPECT_TRUE( tmds->has_string() );
+  EXPECT_FALSE( tmds->has_double() );
+  EXPECT_FALSE( tmds->has_uint64() );
+  EXPECT_EQ( tmds->as_string(), "item data" );
+
+  EXPECT_FALSE( tmdd->has_string() );
+  EXPECT_TRUE( tmdd->has_double() );
+  EXPECT_FALSE( tmdd->has_uint64() );
+  EXPECT_EQ( tmdd->as_double(), 3.14159 );
+  EXPECT_EQ( tmdd->as_string(), "3.14159" );
+
+  EXPECT_FALSE( tmdi->has_string() );
+  EXPECT_FALSE( tmdi->has_double() );
+  EXPECT_TRUE( tmdi->has_uint64() );
+  EXPECT_EQ( tmdi->as_uint64(), 314159 );
+  EXPECT_EQ( tmdi->as_string(), "314159" );
+}
+
+TEST ( metadata, add_metadata )
+{
+  // create item
+  auto* tmdd = new typed_metadata< VITAL_META_PLATFORM_HEADING_ANGLE, double >(
+    std::string( "test double item" ), 3.14159 );
+  auto rmdi = std::make_shared< typed_metadata< VITAL_META_UNIX_TIMESTAMP, uint64_t > >(
+    std::string( "test uint item" ), uint64_t( 314159 ) );
+
+  metadata meta_coll;
+
+  meta_coll.add( NEW_METADATA_ITEM( VITAL_META_METADATA_ORIGIN, std::string( "item data" ) ) );
```

Comment:
BTW, probably this should be `add< VITAL_META_METADATA_ORIGIN >( "item data" )` with #1012. (Note the absence of explicitly constructing a `std::string`! That is expected to work and is worth testing.)